packages/at_secondary_server/lib/src/verb/handler/sync_progressive_verb_handler.dart

```diff
@@ -41,44 +41,56 @@ class SyncProgressiveVerbHandler extends AbstractVerbHandler {
     var atCommitLog = await (AtCommitLogManagerImpl.getInstance()
         .getCommitLog(AtSecondaryServerImpl.getInstance().currentAtSign));
     // Get entries to sync
-    var itr = atCommitLog!.getEntries(
+    var commitEntryIterator = atCommitLog!.getEntries(
         int.parse(verbParams[AT_FROM_COMMIT_SEQUENCE]!) + 1,
         regex: verbParams['regex']);
     // Iterates on all the elements in iterator
     // Loop breaks when the [syncBuffer] reaches the limit.
     // and when syncResponse length equals the [AtSecondaryConfig.syncPageLimit]
-    while (itr.moveNext() &&
+    await populateSyncBuffer(syncBuffer, syncResponse, commitEntryIterator);
+
+    response.data = jsonEncode(syncResponse);
+  }
+
+  @visibleForTesting
+  Future<void> populateSyncBuffer(ByteBuffer syncBuffer,
+      List<dynamic> syncResponse, Iterator<dynamic> commitEntryInterator) async {
+    while (commitEntryInterator.moveNext() &&
         syncResponse.length < AtSecondaryConfig.syncPageLimit) {
       var keyStoreEntry = KeyStoreEntry();
-      keyStoreEntry.key = itr.current.key;
-      keyStoreEntry.commitId = itr.current.value.commitId;
-      keyStoreEntry.operation = itr.current.value.operation;
-      if (itr.current.value.operation != CommitOp.DELETE) {
+      keyStoreEntry.key = commitEntryInterator.current.key;
+      keyStoreEntry.commitId = commitEntryInterator.current.value.commitId;
+      keyStoreEntry.operation = commitEntryInterator.current.value.operation;
+      if (commitEntryInterator.current.value.operation != CommitOp.DELETE) {
         // If commitOperation is update (or) update_all (or) update_meta and key does not
         // exist in keystore, skip the key to sync and continue.
-        if (!keyStore.isKeyExists(itr.current.key)) {
+        if (!keyStore.isKeyExists(commitEntryInterator.current.key)) {
           logger.finer(
-              '${itr.current.key} does not exist in the keystore. skipping the key to sync');
+              '${commitEntryInterator.current.key} does not exist in the keystore. skipping the key to sync');
           continue;
         }
-        var atData = await keyStore.get(itr.current.key);
+        var atData = await keyStore.get(commitEntryInterator.current.key);
         if (atData == null) {
-          logger.info('atData is null for ${itr.current.key}');
+          logger.info('atData is null for ${commitEntryInterator.current.key}');
           continue;
         }
         keyStoreEntry.value = atData.data;
         keyStoreEntry.atMetaData = _populateMetadata(atData);
       }
       // If syncBuffer reaches the limit, break the loop.
-      if (syncBuffer.isOverFlow(utf8.encode(jsonEncode(keyStoreEntry)))) {
+      //[syncBuffer.length() > 0]  s to ensure that atleast one entry
+      // will be synced even though data is overflowing the buffer
+      if (syncResponse.isNotEmpty &&
+          syncBuffer.isOverFlow(utf8.encode(jsonEncode(keyStoreEntry)))) {
```

Comment:
An additional check? No. The check as it is is fine. If you're asking if we should remove that check, the answer is also no - servers must be able to control how much data they are prepared to receive.

Reply:
I agree we should have the check. I'm asking if we need to check if this entry is overflowing the inbound_listener's buffer since we are overriding the syncBuffer in some cases.

Say there is a very large entry which has a size greater than syncBuffer and inboundListener's buffer, but when it's the only entry in the queue we would try to sync it with the current change in this PR which would fail in the later stages. 

Should we put a check, something like (inboundListenerBuffer.isOverflow(syncBuffer)) so that when we are overriding the syncBuffer size, we ensure that it does not exceed the inboundListener's limit?

Reply:
And if you're talking about the server's inbound listener - again, no. 

The point of this PR is to ensure that sync will never get stuck because of this issue. Let's just get this done now and then you can think about how the broader issues if you like